core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/DataProcessor.scala

```diff
@@ -196,20 +196,21 @@ class DataProcessor(
         )
         inputManager.initBatch(channelId, tuples)
         processInputTuple(inputManager.getNextTuple)
-      case EndOfUpstream() =>
-        val channel = this.inputGateway.getChannel(channelId)
-        val portId = channel.getPortId
-
-        this.inputManager.getPort(portId).channels(channelId) = true
-
-        if (inputManager.isPortCompleted(portId)) {
-          inputManager.initBatch(channelId, Array.empty)
-          processInputExhausted()
-          outputManager.outputIterator.appendSpecialTupleToEnd(FinalizePort(portId, input = true))
-        }
-        if (inputManager.getAllPorts.forall(portId => inputManager.isPortCompleted(portId))) {
-          // assuming all the output ports finalize after all input ports are finalized.
-          outputManager.finalizeOutput()
+      case MarkerFrame(marker) =>
+        marker match {
+          case EndOfUpstream() =>
+            this.inputManager.getPort(portId).channels(channelId) = true
+            if (inputManager.isPortCompleted(portId)) {
+              inputManager.initBatch(channelId, Array.empty)
+              processEndOfUpstream(portId.id)
+              outputManager.outputIterator.appendSpecialTupleToEnd(
```

Comment:
They are internal markers inside each operator that are different from those we introduce in this PR. This PR is for markers between operators; we should have a separate PR to deal with internal special tuples if we want to do so.

Reply:
since you refactored Python side's internal markers, can we also align them here? feel free to open another PR for this purpose but I think it should be aligned.

Reply:
I can align it with the python side.

Reply:
Aligned.